### PR TITLE
Relational: Make table reference graph mapping consistent

### LINKED ISSUE #24460: EnumHasFlagTranslator should clone SelectExpression when repeating
Generated SQL
```SQL
SELECT [g].[Nickname], [g].[SquadId], [g].[AssignedCityName], [g].[CityOfBirthName], [g].[Discriminator], [g].[FullName], [g].[HasSoulPatch], [g].[LeaderNickname], [g].[LeaderSquadId], [g].[Rank]
FROM [Gears] AS [g]
WHERE ([g].[Rank] & COALESCE((
    SELECT TOP(1) [g1].[Rank]
    FROM [Gears] AS [g1]
    ORDER BY [g1].[Nickname], [g1].[SquadId]), 0)) = COALESCE((
    SELECT TOP(1) [g1].[Rank]
    FROM [Gears] AS [g1]
    ORDER BY [g1].[Nickname], [g1].[SquadId]), 0)
```
The SelectExpression is visited twice and mutated in place. Current side effect is first visit applies "g0" alias, second one updates to "g1" erasing "g0" from query.

### LINKED ISSUE #24473: Query: Make ColumnExpression.Table consistent in what it refers to
Currently we have to match unwrapping join expression. We should not have to do it.

### PR DESCRIPTION
**Pull request by smitpatel**
Also add a visitor to verify that all table references are proper
Update alias unique-fier to skip already visited expression

Fixes #24473
Fixes #24460

### CODE REVIEW

File: src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs [hunk 1]
```diff
@@ -393,7 +393,8 @@ public void ApplyDistinct()
         /// <summary>
         ///     Adds expressions from projection mapping to projection if not done already.
         /// </summary>
-        public Expression ApplyProjection(Expression shaperExpression, QuerySplittingBehavior querySplittingBehavior)
+        public Expression ApplyProjection(
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs:396]: We already have broken this API in this release. Need resultCardinality for split collections.

File: src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs [hunk 2]
```diff
@@ -3061,6 +3062,18 @@ protected override Expression VisitChildren(ExpressionVisitor visitor)
                 Offset = (SqlExpression?)visitor.Visit(Offset);
                 Limit = (SqlExpression?)visitor.Visit(Limit);
 
+                //var identifier = VisitList(_identifier.Select(e => e.Column).ToList(), inPlace: true, out _)
```
Comment (smitpatel) [inline src/EFCore.Relational/Query/SqlExpressions/SelectExpression.cs:3065]: These are needed for split collection work, but decided to submit that work in separate PR so just left this here commented for now.

File: src/EFCore/Query/Internal/NavigationExpandingExpressionVisitor.cs [hunk 3]
```diff
@@ -944,7 +945,18 @@ private NavigationExpansionExpression ProcessDistinct(NavigationExpansionExpress
                         methodCallExpression.Update(methodCallExpression.Object!, arguments),
                         filterExpression.Parameters);
 
-                    return (result, filterExpression);
+                    if (methodCallExpression.Method.Name.StartsWith("OrderBy", StringComparison.Ordinal)
```
Comment (smitpatel) [inline src/EFCore/Query/Internal/NavigationExpandingExpressionVisitor.cs:948]: string based since we already did stronger check above, so no need of crazy magic here.

File: src/EFCore.Relational/Query/SqlExpressions/SelectExpression.Helper.cs [hunk 4]
```diff
@@ -365,6 +369,14 @@ public void UpdateTableReference(SelectExpression oldSelect, SelectExpression ne
                 }
             }
 
+            internal void Verify(SelectExpression selectExpression)
+            {
+                if (!ReferenceEquals(selectExpression, _selectExpression))
+                {
+                    throw new InvalidOperationException("Dangling TableReferenceExpression.");
```
Comment (maumar) [inline src/EFCore.Relational/Query/SqlExpressions/SelectExpression.Helper.cs:376]: Don't forget to fix those messages (in a future checkin)

File: src/EFCore.Relational/Query/SqlExpressions/SelectExpression.Helper.cs [hunk 5]
```diff
@@ -584,5 +604,125 @@ public ClientProjectionRemappingExpressionVisitor(object[] clientProjectionIndex
                 return base.Visit(expression);
             }
         }
+
+        private sealed class SelectExpressionVerifyingExpressionVisitor : ExpressionVisitor
```
Comment (maumar) [inline src/EFCore.Relational/Query/SqlExpressions/SelectExpression.Helper.cs:608]: this visitor is not used anywhere - purely a debugging tool? (i.e. we don't want to run always in debug, like we do with table aliases?

Comment (smitpatel) [reply src/EFCore.Relational/Query/SqlExpressions/SelectExpression.Helper.cs:608]: Yes, it is for debugging. Split query introduces clone method on select expression which was causing a lot of errors (for obvious reason), so I added this. I will see if there is a place where I can add this. It is somewhat more complicated than table aliases as this is more stricter form of graph consistency, so it cannot be run during most point when we are mutating SelectExpression.

**Review by maumar — APPROVED** [review]
